Azure.ResourceManager.Core/Resources/Location.cs

```diff
@@ -2,98 +2,458 @@
+        /// </summary>
+        /// <param name="other">Location object or name as a string.</param>
+        /// <returns>-1 for less than, 0 for equals, 1 for greater than.</returns>
         public int CompareTo(Location other)
         {
             if (ReferenceEquals(other, null))
             {
                 return 1;
             }
 
-            return string.Compare(Name, other.Name);
+            return Name.CompareTo(other.Name);
         }
 
-        public int CompareTo(string other)
+        private static string NormalizationUtility(string value)
         {
-            if (ReferenceEquals(other, null))
+            if (string.IsNullOrEmpty(value))
             {
-                return 1;
+                return value;
             }
 
-            return Name.CompareTo(other);
-        }
+            var sb = new StringBuilder(value.Length);
+            for (var index = 0; index < value.Length; ++index)
+            {
+                var c = value[index];
+                if (char.IsLetterOrDigit(c))
+                {
+                    sb.Append(c);
+                }
+            }
 
-        public bool Equals(Location other)
-        {
-            return CanonicalName == other.CanonicalName;
+            return sb.ToString().ToUpperInvariant();
         }
 
-        public bool Equals(string other)
+        private static NameType DetectNameType(string location)
         {
-            return CanonicalName == GetCanonicalName(other);
+            if (Regex.IsMatch(location, CanonicalPattern))
+            {
+                return NameType.CanonicalName;
+            }
+            else if (Regex.IsMatch(location, DisplayPattern))
+            {
+                return NameType.DisplayName;
+            }
+            else
+            {
+                return NameType.Name;
+            }
         }
 
-        public override string ToString()
+        private static string GetCanonicalNameFromDisplayName(string name)
         {
-            return DisplayName;
+            return Regex.Replace(name.ToLower(), @" ", "-");
         }
 
-        private static string GetCanonicalName(string name)
+        private static string GetDisplayNameFromCanonicalName(string name)
         {
-            return name;
+            char[] chName = name.ToCharArray();
+            chName[0] = char.ToUpper(chName[0]);
+
+            for (int i = 0; i < chName.Length - 1; i++)
+            {
+                if (chName[i] == '-')
+                {
+                    chName[i] = ' ';
+                    chName[i + 1] = char.ToUpper(chName[i + 1]);
+                }
+            }
+
+            return new string(chName);
         }
 
-        private static string GetDisplayName(string name)
+        private static string GetDefaultNameFromCanonicalName(string name)
         {
-            return name;
+            return Regex.Replace(name, RegexDash, string.Empty);
```

Comment:
Can we see whats faster here
Regex.Replace(name, RegexDash, string.Empty)
vs
name.Replace(" ", string.Empty)

Reply:
Done, name.Replace definetly runs faster than Regex.Replace.
![image](https://user-images.githubusercontent.com/17486462/102661168-2da62c00-4131-11eb-8e2c-9bb338f87c36.png)

I'll replace all the Regex.Replace with the string Replace method.